feat: show longer dbt descriptions for fields, including markdown support

filp opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

Closes: #1174 

### Description:

Adds an improved UX for viewing dbt field descriptions:

<img width="478" alt="Screenshot 2024-02-20 at 22 15 58" src="https://github.com/lightdash/lightdash/assets/382538/c693492c-c47a-498f-80af-e955b06e39eb">
<img width="420" alt="Screenshot 2024-02-20 at 22 16 17" src="https://github.com/lightdash/lightdash/assets/382538/c60091e1-e419-43d4-b412-f996627c0688">
<img width="727" alt="Screenshot 2024-02-20 at 22 16 25" src="https://github.com/lightdash/lightdash/assets/382538/f8997942-198c-40ec-b115-ffbc78d25bc1">
<img width="700" alt="Screenshot 2024-02-20 at 22 17 16" src="https://github.com/lightdash/lightdash/assets/382538/68e25936-b58f-4720-b728-5bc47f396769">
<img width="713" alt="Screenshot 2024-02-20 at 22 17 09" src="https://github.com/lightdash/lightdash/assets/382538/e2149f29-1937-47fb-9f05-72e09ad6ce07">


### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

hamzahc1 commented:
@filp what's left to do to open this PR for review?

filp commented:
@hamzahc1 there was a quirk with rendering modals from within the field tree on the left that I had to rework, and some design feedback to integrate. Opened it up for review now!

filp commented:
Currently still includes one quirk I couldn't find a satisfactory solution for: opening the description modal from the hovercard doesn't immediately close the hovercard even though the modal is already visible, because the user's hover state is still keeping the card open.

Will take another stab at a solution, but IMO doesn't block reviewing the overall change.

rephus commented on packages/frontend/src/components/Explorer/ExploreTree/TableTree/Tree/TreeSingleNode.tsx:
````diff
@@ -10,23 +10,128 @@ import {
+    const truncatedLines = lines
+        .slice(0, 4)
+        .reduce<string[]>((truncated, line, idx) => {
+            if (idx === 0 || (line.at(0) !== '|' && !line.startsWith('```'))) {
````
so if the description starts with a block of code we will not show anything, right ?  I guess it kind of makes sense...

rephus reviewed (approved):
Tested and review, it look good to me, nice job!

Tip for make reviews easier: it would be nice if you can include a `before` and `after` screenshot with the change. 

I can see `last name` on render has a long description, but I don't see this change on jaffle shop in this PR. I think it would be nice to push that to the repo as well (or a shorter version), so we have something to test this again in the future.  

> Currently still includes one quirk I couldn't find a satisfactory solution for: opening the description modal from the hovercard doesn't immediately close the hovercard even though the modal is already visible, because the user's hover state is still keeping the card open.

I also noticed this, if you can fix it great, but if not, I think it would be ok to ship it as it is.

filp commented:
Thank you @rephus! I made a couple of corrections based on your feedback:

- Tweaked the description for 'Order status' to include a bit more content to test this feature.
- Fixed a pending issue with markdown headings auto-linking
- Fixed the weird interaction between the hover-card and the modal by using a controlled Popover instead
- Simplified/improved the on-hover preview by rendering the full description as markdown, and instead clamping the visible amount in the hover card (so there's less magic when deciding what to render).

See a quick gif here:

![Kapture 2024-02-23 at 14 24 01](https://github.com/lightdash/lightdash/assets/382538/21928206-d1b5-437d-8e3b-f188b6fc60aa)

Would love it if you could take another quick look and share you opinion!

rephus reviewed (approved):
Everything look good, it looks even better now. This hasn't changed though. Not sure if you are planning to do it 

https://github.com/lightdash/lightdash/pull/9055#issuecomment-1961214743

filp commented:
Checked the failing e2e test locally, doesn't appear to be related.
